docs/generic-templates.md

```diff
@@ -0,0 +1,173 @@
+```haskell
+template T a with ...
+````
+All types `R`, `U_i` and `V_j` have the type variable `a` in scope.
+
+An instantiation of this generic template with an actual type `S` in place of type parameter `a` can be declared through
+```haskell
+template instance I = T S
+```
+The `I` is the name of the underlying DAML-LF type and template that will be created for this instantiation. This type will also exist in DAML but is not intended to be used there.
+
+Extending this syntax to more than one type parameter is trivial.
+
+## Underlying type classes
+
+We split the current `Template` type class into three type classes serving the following purpose:
+1. `LfTemplate` indicates an "actual template" which will cause a template definition in DAML-LF. Our example above will yield `instance LfTemplate I`. The `LfTemplate` type class serves the same purpose as the current `Template` type class but is not intended to be used in DAML. Instances of `LfTemplate` are generated for generic template instantiations and non-generic template definitions.
+1. `Template` indicates a "logical template" which exists only in DAML but not in DAML-LF. Our example above will yield `instance D a => Template (T a)`. Instances of `Template` are generated for all template definitions, generic and non-generic ones.
+1. `Creatable` links an instance of `Template` with an instance of `LfTemplate` (at the value level). Our example above will yield `instance Creatable (T S)`. Instances of `Creatable` are generated for generic template instantiations and non-generic template definitions.
+
+The actual definitions of the three type classes are as follows:
+```haskell
+class LfTemplate x where
+  lfSignatory : t -> [Party]
+  lfCreate : t -> Update (ContractId t)
+  lfCreate = magic @"create"
+  lfFetch : ContractId t -> Update t
+  lfFetch = magic @"fetch"
+
+class Template t where
+  signatory : t -> [Party]
+
+class Template t => Creatable t where
+  create : t -> Update (ContractId t)
+  fetch : ContractId t -> Update t
+```
+
+Similarly, we split the `Choice` type class into `LfChoice`, `Choice` and `Exercisable`:
+```haskell
+class LfTemplate t => LfChoice t c r | t c -> r where
+  lfController : t -> c -> [Party]
+  lfAction : ContractId t -> t -> c -> Update r
+  lfExercise : ContractId t -> c -> Update r
+  lfExercise = magic @"exercise"
+
+class Template t => Choice t c r | t c -> r where
+  controller : t -> c -> [Party]
+  action : ContractId t -> t -> c -> Update r
+
+class (Creatable t, Choice t c r) => Exercisable t c r | t c -> r where
+  exercise : ContractId t -> c -> Update r
+```
+
+## Desugaring generic template definitions
+
+Desugarting the generic template definition of `T` from above will generate the following data definitions type class instances:
+```haskell
+data T a = T with f_1 : U_1; ...
+
+instance D a => Template (T a)
+  signatory this@T{..} = sign
```

Comment:
Suggest `templateSignatory` (see next comment).

Reply:
I'd prefer leaving this as `signatory` because it allows for writing `signatory x` to get the signatories of some contract instance `x`. That we have `choiceController` is a workaround for the special properties of layout introducing keywords. In fact, I'd be very happy if we found a way to have a function called `controller` in the future again.

Reply:
Ok, scrap the syntactic obfuscation idea. Indeed, `controller` can be recovered if and when we drop support for non-flexible choice syntax and `controller` as layout inducing.